### PR TITLE
Add stylecop

### PR DESCRIPTION
**Pull request by jbradford**
Added codestyles + refactor.
Added stylecop using CR.CodeStyle.CSharp.Full.
Refactored code to remove duplication.
Used modern c# features in existing code.

### CODE REVIEW

File: src/Persistence.RavenDB/RavenDBViewModelHelper.cs [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    {
+        /// <summary>
+        /// Helper method used to generate the ID of the RavenDB document that a view model of type TEntity with specified key should be stored.
+        /// </summary>
+        /// <typeparam name="TEntity">The type of the View Model</typeparam>
+        /// <param name="key">The key of the view model</param>
+        /// <returns>The ID of the RavenDB document.</returns>
+        internal static string MakeID<TEntity>(string key)
+        {
+            return typeof(TEntity).FullName + "/" + key;
```
Comment (ghost) [inline src/Persistence.RavenDB/RavenDBViewModelHelper.cs]: String interpolation would be a _little_ nicer, especially with an expression bodied property.

File: src/core/IViewModelWriter.cs [hunk 2]
```diff
@@ -1,18 +1,74 @@
...
     {
-        void Add<TEntity>(string key, TEntity entity) where TEntity : class;
+        /// <summary>
+        /// Attempts to add the specified view model of type TEntity to the view model storage.
+        /// </summary>
+        /// <typeparam name="TEntity">The type of the view model to store.</typeparam>
+        /// <param name="key">The key of the view model being stored.</param>
+        /// <param name="entity">The view model of type TEntity that is being stored.</param>
+        /// <exception cref="DuplicateKeyException">Exception will be thrown when the given key for a
+        /// view model of type TEntity is already stored in the view model storage for this </exception>
```
Comment (ghost) [inline src/core/IViewModelWriter.cs]: Whitespace & no .

File: src/core/IViewModelWriter.cs [hunk 3]
```diff
@@ -1,18 +1,74 @@
...
+        /// Attempts to update the specified view model of type TEntity in the view model storage
+        /// using the provided delegate.
+        /// </summary>
+        /// <typeparam name="TEntity">The type of the view model to update.</typeparam>
+        /// <param name="key">The key of the view model being updated.</param>
+        /// <param name="update">The update delegate that should be performed on the view model.</param>
+        /// <exception cref="ArgumentNullException">Exception will be thrown whenever null is passed in for either of the parameters.</exception>
+        /// <exception cref="ArgumentException">Exception will be thrown whenever an empty string is passed in for parameter key.</exception>
+        /// <exception cref="EntityNotFoundException">Exception will be thrown when attempting to update
+        /// a key that is not present in the view model storage</exception>
```
Comment (ghost) [inline src/core/IViewModelWriter.cs]: No .

File: src/core/IViewModelWriter.cs [hunk 4]
```diff
@@ -1,18 +1,74 @@
...
+        /// <summary>
+        /// Attempts to delete the specified view model of type TEntity from the view model storage
+        /// using the provided delegate.
+        /// </summary>
+        /// <typeparam name="TEntity">The type of the view model to delete.</typeparam>
+        /// <param name="key">The key of the view model being deleted.</param>
+        /// <exception cref="ArgumentNullException">Exception will be thrown whenever null is passed in for parameter key.</exception>
+        /// <exception cref="ArgumentException">Exception will be thrown whenever an empty string is passed in for parameter key.</exception>
+        /// <exception cref="EntityNotFoundException">Exception will be thrown when attempting to delete
+        /// a key that is not present in the view model storage</exception>
```
Comment (ghost) [inline src/core/IViewModelWriter.cs]: No .

**Review by ghost — DISMISSED** [review]
Nothing significant; couple of minor docs notes.

Resharper has a significant number of good suggestions such as using nameof in places & making certain properties & fields that could be readonly readonly, and while it's not mandated by Stylecop, it might be worth going through and cleaning them up.

**Review by lparkermg — DISMISSED** [review]
LGTM

File: src/Persistence.ApplicationState/Persistence.ApplicationState.csproj [hunk 5]
```diff
@@ -51,5 +59,21 @@
       <Name>Core</Name>
     </ProjectReference>
   </ItemGroup>
+  <ItemGroup>
+    <AdditionalFiles Include="..\..\..\..\Users\isaac.beizsley\.nuget\packages\CR.CodeStyle.CSharp.Full\1.0.1\stylecop.json">
```
Comment (jbradford) [inline src/Persistence.ApplicationState/Persistence.ApplicationState.csproj]: This looks wrong...?

File: src/Persistence.RavenDB/RavenDBViewModelHelper.cs [hunk 6]
```diff
@@ -7,17 +7,14 @@ namespace CR.ViewModels.Persistence.RavenDB
     /// <summary>
     /// Helper class used for code shared between <see cref="RavenDBViewModelReader"/> and <see cref="RavenDBViewModelWriter"/>
     /// </summary>
-    internal static class RavenDBViewModelHelper
+    internal static class RavenDbViewModelHelper
```
Comment (jbradford) [inline src/Persistence.RavenDB/RavenDBViewModelHelper.cs]: 2 letter acronyms (e.g. DB) left purposefully that way based on MS guidelines.

https://docs.microsoft.com/en-us/previous-versions/dotnet/netframework-1.1/141e06ef(v=vs.71)

File: src/Persistence.RavenDB/RavenDBViewModelWriter.cs [hunk 7]
```diff
@@ -26,10 +28,7 @@ public class RavenDBViewModelWriter : IViewModelWriter
         /// Initializes a new instance of the <see cref="RavenDBViewModelWriter"/> class.
```
Comment (jbradford) [inline src/Persistence.RavenDB/RavenDBViewModelWriter.cs]: We should consistently use the `T:CR.ViewModels.Core.IViewModelWriter` format if we're going to use it in the docs.

File: src/tests/packages.config [hunk 8]
```diff
@@ -1,10 +1,12 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <packages>
+  <package id="CR.CodeStyle.CSharp.Full" version="1.0.1" targetFramework="net471" developmentDependency="true" />
```
Comment (jbradford) [inline src/tests/packages.config:3]: Code Styles unnecessary in test projects (although no need to remove them now they've been added.).

**Review by jbradford — COMMENTED** [review]
More comments.

File: src/Persistence.RavenDB/RavenDBViewModelHelper.cs [hunk 9]
```diff
@@ -0,0 +1,20 @@
+﻿// <copyright file="RavenDBViewModelHelper.cs" company="Cognisant">
+// Copyright (c) Cognisant. All rights reserved.
+// </copyright>
+
+namespace CR.ViewModels.Persistence.RavenDB
+{
+    /// <summary>
+    /// Helper class used for code shared between <see cref="RavenDBViewModelReader"/> and <see cref="RavenDBViewModelWriter"/>.
+    /// </summary>
+    internal static class RavenDbViewModelHelper
```
Comment (jbradford) [inline src/Persistence.RavenDB/RavenDBViewModelHelper.cs]: Comments seems to have disappeared on this one. 2 Letter acronyms should be left categorised based on MS Guidelines https://docs.microsoft.com/en-us/previous-versions/dotnet/netframework-1.1/141e06ef(v=vs.71)

File: src/Persistence.ApplicationState/ApplicationStateViewModelReader.cs [hunk 10]
```diff
@@ -1,45 +1,48 @@
...
+    /// <inheritdoc />
+    /// <summary>
+    /// A implementation of the <see cref="T:CR.ViewModels.Core.IViewModelReader" /> interface that uses
+    /// HttpApplicationState from System.Web to retrieve view models.
+    /// </summary>
     public class ApplicationStateViewModelReader : IViewModelReader
     {
-        private HttpApplicationStateBase AppState { get; set; }
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ApplicationStateViewModelReader"/> class, using a provided <see cref="HttpApplicationStateBase"/> to access view models.
```
Comment (jbradford) [inline src/Persistence.ApplicationState/ApplicationStateViewModelReader.cs:20]: Another lost comment.
We should consistently use the `T:CR.ViewModels.Core.IViewModelWriter` format if we're going to use it in the docs.

File: src/Persistence.RavenDB/RavenDBViewModelHelper.cs [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+    // ReSharper disable once InconsistentNaming
+    internal static class RavenDBViewModelHelper
+    {
+        /// <summary>
+        /// Helper method used to generate the ID of the RavenDB document that a view model of type TEntity with specified key should be stored.
+        /// </summary>
+        /// <typeparam name="TEntity">The type of the View Model.</typeparam>
+        /// <param name="key">The key of the view model.</param>
+        /// <returns>The ID of the RavenDB document.</returns>
+        internal static string MakeId<TEntity>(string key) => $"{typeof(TEntity).FullName}/{key}";
```
Comment (jbradford) [inline src/Persistence.RavenDB/RavenDBViewModelHelper.cs]: Should probably fix ID as well :p

**Comment by jbradford** [comment]
LGTM, I can't review though @rcknight

**Pull request merged**